pkg/experiment/symbolizer/symbolizer.go

```diff
@@ -0,0 +1,448 @@
+	var framesBuf []lidia.SourceInfoFrame
+
+	resolveStart := time.Now()
+	defer func() {
+		s.metrics.debugSymbolResolution.WithLabelValues(statusSuccess).Observe(time.Since(resolveStart).Seconds())
+	}()
+
+	for _, loc := range req.locations {
+		frames, err := table.Lookup(framesBuf, loc.address)
+		if err != nil {
+			loc.lines = s.createNotFoundSymbols(req.binaryName, loc)
+			continue
+		}
+
+		if len(frames) == 0 {
+			loc.lines = s.createNotFoundSymbols(req.binaryName, loc)
+			continue
+		}
+
+		loc.lines = frames
+	}
+}
+
+func (s *Symbolizer) getLidiaBytes(ctx context.Context, buildID string) ([]byte, error) {
+	lidiaBytes, err := s.fetchLidiaFromObjectStore(ctx, buildID)
+	if err == nil {
+		return lidiaBytes, nil
+	}
+
+	level.Error(s.logger).Log(
+		"msg", "failed to get from object store",
+		"build_id", buildID,
+		"error", err,
+	)
+
+	lidiaBytes, err = s.fetchLidiaFromDebuginfod(ctx, buildID)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := s.store.Put(ctx, buildID, bytes.NewReader(lidiaBytes)); err != nil {
+		level.Warn(s.logger).Log("msg", "Failed to store debug info in objstore", "buildID", buildID, "err", err)
+	}
+
+	return lidiaBytes, nil
+}
+
+// fetchLidiaFromObjectStore retrieves Lidia data from the object store
+func (s *Symbolizer) fetchLidiaFromObjectStore(ctx context.Context, buildID string) ([]byte, error) {
+	objstoreReader, err := s.store.Get(ctx, buildID)
+	if err != nil {
+		return nil, err
+	}
+	defer objstoreReader.Close()
+
+	data, err := io.ReadAll(objstoreReader)
+	if err != nil {
+		return nil, fmt.Errorf("read content: %w", err)
+	}
+
+	return data, nil
+}
+
+// fetchLidiaFromDebuginfod fetches debug info from debuginfod and converts to Lidia format
+func (s *Symbolizer) fetchLidiaFromDebuginfod(ctx context.Context, buildID string) ([]byte, error) {
+	debuginfodStart := time.Now()
+	status := statusSuccess
+
+	debugReader, err := s.fetchFromDebuginfod(ctx, buildID)
+	if err != nil {
+		var bnfErr buildIDNotFoundError
+		if errors.As(err, &bnfErr) {
+			return nil, err
+		}
+		return nil, err
+	}
+	s.metrics.debuginfodRequestDuration.WithLabelValues(status).Observe(time.Since(debuginfodStart).Seconds())
+	defer debugReader.Close()
+
+	elfData, err := io.ReadAll(debugReader)
+	if err != nil {
+		return nil, fmt.Errorf("read debuginfod data: %w", err)
+	}
+
+	lidiaData, err := s.processELFData(elfData)
+	if err != nil {
+		return nil, err
+	}
+
+	return lidiaData, nil
+}
+
+func (s *Symbolizer) fetchFromDebuginfod(ctx context.Context, buildID string) (io.ReadCloser, error) {
+	debuginfodStart := time.Now()
+
+	debugReader, err := s.client.FetchDebuginfo(ctx, buildID)
+	if err != nil {
+		var bnfErr buildIDNotFoundError
+		statusCode, isHTTPError := isHTTPStatusError(err)
+
+		if errors.As(err, &bnfErr) || (isHTTPError && statusCode == http.StatusNotFound) {
+			s.metrics.debuginfodRequestDuration.WithLabelValues(statusErrorNotFound).Observe(time.Since(debuginfodStart).Seconds())
```

Comment:
please take another look at this metric. Something is off here. it looks like error statuses are accounted multiple times, in the DebugInfodClient and in the symbolizer. Can we remove all the usages of debuginfod metrics from the symbolizer?